File: backend/app/schemas/geojson.py
from typing import Any, Literal

from pydantic import BaseModel, Field


class Feature(BaseModel):
    type: Literal["Feature"] = "Feature"
    geometry: dict[str, Any]
    properties: dict[str, Any] = Field(default_factory=dict)


class FeatureCollection(BaseModel):
    type: Literal["FeatureCollection"] = "FeatureCollection"
    features: list[Feature] = Field(default_factory=list)
    crs: dict[str, Any] | None = Field(
        default={"type": "name", "properties": {"name": "EPSG:4326"}}
    )

File: frontend/src/types/geojson.ts
import type { Feature, FeatureCollection as BaseFeatureCollection, LineString } from 'geojson'

export type GeoJsonProperties = Record<string, string | number | boolean | null | string[]>

export type GeoJsonFeature = Feature<LineString, GeoJsonProperties>

export interface FeatureCollection extends BaseFeatureCollection<LineString, GeoJsonProperties> {
  crs?: {
    type: 'name'
    properties: {
      name: 'EPSG:4326' | string
    }
  }
}
